README.rst
Requests-Hawk
#############

|pypi| |travis|

.. |travis| image:: https://travis-ci.org/mozilla-services/requests-hawk.png
    :target: https://travis-ci.org/mozilla-services/requests-hawk

.. |pypi| image:: https://img.shields.io/pypi/v/requests-hawk.svg
    :target: https://pypi.python.org/pypi/requests-hawk


This project allows you to use `the python requests library
<http://python-requests.org/>`_ with `the hawk authentication
<https://github.com/hueniverse/hawk>`_ mechanism.

Hawk itself does not provide any mechanism for obtaining or transmitting the
set of shared credentials required, but this project proposes a scheme we use
across mozilla services projects.

Great, how can I use it?
========================

First, you'll need to install it:

.. code-block:: bash

    pip install requests-hawk

Then, in your project, if you know the `id` and `key`, you can use:

.. code-block:: python

    import requests
    from requests_hawk import HawkAuth

    hawk_auth = HawkAuth(id='my-hawk-id', key='my-hawk-secret-key')
    requests.post("https://example.com/url", auth=hawk_auth)

Or if you need to derive them from the hawk session token, instead use:

.. code-block:: python

    import requests
    from requests_hawk import HawkAuth

    hawk_auth = HawkAuth(
        hawk_session=resp.headers['hawk-session-token'],
        server_url=self.server_url
    )
    requests.post("/url", auth=hawk_auth)

In the second example, the ``server_url`` parameter to ``HawkAuth`` was used to
provide a default host name, to avoid having to repeat it for each request.

If you wish to override the default algorithm of ``sha256``, pass the desired
algorithm name using the optional ``algorithm`` parameter.

Note: The ``credentials`` parameter has been removed. Instead pass ``id`` and
``key`` separately (as above), or pass the existing dict as ``**credentials``.

Integration with httpie
=======================

`Httpie <https://github.com/jakubroztocil/httpie>`_ is a tool which lets you do
requests to a distant server in a nice and easy way. Under the hood, ``httpie``
uses the requests library. We've made it simple for you to plug hawk with it.

If you know the id and key, use it like that:

.. code-block:: bash

   http POST localhost:5000/registration\
   --auth-type=hawk --auth='id:key'

Or, if you want to use the hawk session token, you can do as follows:

.. code-block:: bash

   http POST localhost:5000/registration\
   --auth-type=hawk --auth='c0d8cd2ec579a3599bef60f060412f01f5dc46f90465f42b5c47467481315f51:'

Take care, don't forget to add the extra ``:`` at the end of the hawk session
token for it to be considered like so.

How are the shared credentials shared?
======================================

Okay, on to the actual details.

The server gives you a session token, that you'll need to derive to get the
hawk credentials.

Do an HKDF derivation on the given session token. You'll need to use the
following parameters:

.. code-block:: python

    key_material = HKDF(hawk_session, '', 'identity.mozilla.com/picl/v1/sessionToken', 32*2)

The key material you'll get out of the HKDF needs to be separated into two
parts, the first 32 hex characters are the ``hawk id``, and the next 32 ones are the
``hawk key``:

.. code-block:: python

    credentials = {
        'id': keyMaterial[0:32]
        'key': keyMaterial[32:64]
        'algorithm': 'sha256'
    }

Run tests
=========

To run test, you can use tox:

.. code-block:: bash

    tox


CHANGES.txt
CHANGELOG
=========

1.2.1 (2023-05-03)
------------------

- Add support for ext external data string (#34)
  (thanks @jtmaclachlan)

1.2.0 (unreleased)
------------------

- Nothing changed yet.


1.1.1 (2021-06-04)
------------------

- Handle cases where Content-Type is defined as bytes rather than string. (#25)
- Allow for app mohawk sender parameter configuration


1.1.0 (2020-12-16)
------------------

- Allow to skip hashing request and response bodies with `always_hash_content`.


1.0.1 (2020-01-20)
------------------

- Add Python 3 support. (#22)


1.0.0 (2015-12-15)
------------------

- Simplified API for using HawkAuth when the id and key are known. (#8)
- Added support for overriding the default algorithm (sha256) when deriving
  credentials from the hawk session token, via a new ``algorithm`` parameter.

See the README for migration advice if you use the ``credentials`` parameter.


0.2.1 (2015-10-14)
------------------

- Make sure the requests json parameter is handled properly. (#7)


0.2.0 (2015-05-19)
------------------

- Fix encoding error in setup.py with Python 3.4
- Add a configuration parameter in order to be able to set the
  timestamp to use.


0.1.2 (2014-08-13)
------------------

- Add Python3 support


0.1.1 (2014-07-21)
------------------

- First version


requests_hawk/tests/test_hawkauth.py
import codecs
import unittest
from requests import Request
from requests_hawk import HawkAuth


class TestHawkAuth(unittest.TestCase):

    def test_hawkauth_errors_when_id_and_key_and_hawk_session_passed(self):
        self.assertRaises(AttributeError, HawkAuth,
                          id='test', key='test', hawk_session="test")

    def test_hawkauth_errors_when_id_and_hawk_session_passed(self):
        self.assertRaises(AttributeError, HawkAuth,
                          id='test', hawk_session="test")

    def test_hawkauth_errors_when_key_and_hawk_session_passed(self):
        self.assertRaises(AttributeError, HawkAuth,
                          key='test', hawk_session="test")

    def test_hawkauth_errors_when_only_id_passed(self):
        self.assertRaises(AttributeError, HawkAuth, id='test')

    def test_hawkauth_errors_when_only_key_passed(self):
        self.assertRaises(AttributeError, HawkAuth, key='test')

    def test_hawkauth_errors_when_credentials_passed(self):
        self.assertRaises(AttributeError, HawkAuth, credentials={})

    def test_hawkauth_errors_when_no_auth_is_set(self):
        self.assertRaises(AttributeError, HawkAuth)

    def test_hawk_auth_supports_credentials_as_parameters(self):
        auth = HawkAuth(id='test_id', key='test_key', _timestamp=1431698426)
        request = Request('PUT', 'http://www.example.com',
                          json={"foo": "bar"}, auth=auth)
        r = request.prepare()
        auth_header = r.headers['Authorization']
        self.assertTrue('id="test_id"' in auth_header, "ID doesn't match")
        self.assertTrue('hash="lby/vvNtW9J/ZI39Nnfcrl2ESgx8JIIZ/SaIvfd3iaQ="'
                        in auth_header,
                        "Hash doesn't match")
        self.assertTrue('ts="1431698426"' in auth_header,
                        "Timestamp doesn't match")
        self.assertEqual(r.body, b'{"foo": "bar"}')

    def test_overriding_credentials_algorithm(self):
        auth = HawkAuth(id='test_id', key='test_key', algorithm='sha1')
        self.assertEqual(auth.credentials['algorithm'], 'sha1')

    def test_key_non_hex_values_throws(self):
        self.assertRaises(TypeError, HawkAuth, hawk_session="test")

    def test_credentials_are_derived_from_session(self):
        auth = HawkAuth(hawk_session=codecs.encode(b"hello", "hex_codec"))
        self.assertEqual(auth.credentials, {
            'id': b'15064c77e946608226a9c2d8da61ac5e0e85f325334965c68a3f47e809'
                  b'1f8412',
            'key': b'cb3829c6d6fe3f58609d58f09818295dfbdf45803ec50b8d66c4132f7'
                   b'ad14aa0',
            'algorithm': 'sha256'
        })

    def test_overriding_session_algorithm(self):
        auth = HawkAuth(hawk_session=codecs.encode(b"hello", "hex_codec"),
                        algorithm='sha1')
        self.assertEqual(auth.credentials['algorithm'], 'sha1')

    def test_server_url_is_parsed(self):
        auth = HawkAuth(hawk_session=codecs.encode(b"hello", "hex_codec"),
                        server_url="http://localhost:5000")
        self.assertEquals(auth.host, "localhost:5000")

    def test_hawk_auth_can_handle_a_timestamp_argument(self):
        auth = HawkAuth(hawk_session=codecs.encode(b"hello", "hex_codec"),
                        _timestamp=1431698847)

        request = Request('PUT', 'http://www.example.com',
                          {'Content-Type': 'application/json'},
                          data='{"foo": "bar"}', auth=auth)

        r = request.prepare()
        self.assertTrue('ts="1431698847"' in r.headers['Authorization'],
                        "Timestamp doesn't match")
        self.assertEqual(r.body, '{"foo": "bar"}')

    def test_hawk_auth_is_called_when_json_present(self):
        auth = HawkAuth(hawk_session=codecs.encode(b"hello", "hex_codec"),
                        _timestamp=1431698847)

        request = Request('PUT', 'http://www.example.com',
                          json={"foo": "bar"}, auth=auth)
        r = request.prepare()

        self.assertTrue('ts="1431698847"' in r.headers['Authorization'],
                        "Timestamp doesn't match")
        self.assertEqual(r.body, b'{"foo": "bar"}')

    def test_hawk_auth_supports_empty_body(self):
        auth = HawkAuth(id='test_id', key='test_key', always_hash_content=False)
        request = Request('GET', 'http://www.example.com', auth=auth)
        r = request.prepare()
        self.assertNotIn('hash="', r.headers['Authorization'])

    def test_hawk_auth_supports_binary_content_type(self):
        headers = {'Content-Type': b'application/json'}
        auth = HawkAuth(id='test_id', key='test_key')
        request = Request('POST', 'https://example.com', auth=auth,
                          headers=headers, data=b"data")
        r = request.prepare()
        auth_header = r.headers['Authorization']
        self.assertTrue('id="test_id"' in auth_header, "ID doesn't match")


requests_hawk/__init__.py
import binascii
import codecs
import hashlib
import hmac
import math
from six.moves import xrange
from six.moves.urllib.parse import urlparse
from six import text_type

import mohawk
from mohawk.base import EmptyValue
from requests.auth import AuthBase


class HawkAuth(AuthBase):
    """Handles authentication using Hawk.

    :param hawk_session:
      The hawk session, from the server, encoded as hexadecimal.
      You don't need to set this parameter if you already know the hawk
      credentials (Optional).

    :param id:
      The hawk id string to use for authentication (Optional).

    :param key:
      A string containing the hawk secret key (Optional).

    :param algorithm:
      A string containing the name of the algorithm to be used.
      (Optional, defaults to 'sha256').

    :param server_url:
      The url of the server, this is useful for hawk when signing the requests.
      In case this is omitted, fallbacks to the value of the "Host" header of
      the request (Optional).

    :param ext:
      A string of arbitrary data to be sent along with the request (Optional).

    Note that the `hawk_session` and `id` parameters are mutually exclusive.
    You should use either `hawk_session` or both `id` and 'key'.
    """
    def __init__(self, hawk_session=None, id=None, key=None, algorithm='sha256',
                 credentials=None, server_url=None, _timestamp=None,
                 always_hash_content=True, ext=None, app=None):
        if credentials is not None:
            raise AttributeError("The 'credentials' param has been removed. "
                                 "Pass 'id' and 'key' instead, or '**credentials_dict'.")

        if (hawk_session and (id or key)
                or not hawk_session and not (id and key)):
            raise AttributeError("You should pass either 'hawk_session' "
                                 "or both 'id' and 'key'.")

        if hawk_session:
            try:
                hawk_session = codecs.decode(hawk_session, 'hex_codec')
            except binascii.Error as e:
                raise TypeError(e)
            keyInfo = 'identity.mozilla.com/picl/v1/sessionToken'
            keyMaterial = HKDF(hawk_session, "", keyInfo, 32*2)
            id = codecs.encode(keyMaterial[:32], "hex_codec")
            key = codecs.encode(keyMaterial[32:64], "hex_codec")

        self.credentials = {
            'id': id,
            'key': key,
            'algorithm': algorithm
        }
        self._timestamp = _timestamp
        self.host = urlparse(server_url).netloc if server_url else None
        self.always_hash_content = always_hash_content
        self.ext = ext
        self.app = app

    def __call__(self, r):
        if self.host is not None:
            r.headers['Host'] = self.host

        content_type = r.headers.get("Content-Type") or ""
        if not isinstance(content_type, text_type):
            content_type = content_type.decode("utf-8")

        sender = mohawk.Sender(
            self.credentials,
            r.url,
            r.method,
            content=r.body or EmptyValue,
            content_type=content_type or EmptyValue,
            always_hash_content=self.always_hash_content,
            _timestamp=self._timestamp,
            ext=self.ext,
            app=self.app
        )

        r.headers['Authorization'] = sender.request_header
        return r


def HKDF_extract(salt, IKM, hashmod=hashlib.sha256):
    """HKDF-Extract; see RFC-5869 for the details."""
    if salt is None:
        salt = b"\x00" * hashmod().digest_size
    if isinstance(salt, text_type):
        salt = salt.encode("utf-8")
    return hmac.new(salt, IKM, hashmod).digest()


def HKDF_expand(PRK, info, L, hashmod=hashlib.sha256):
    """HKDF-Expand; see RFC-5869 for the details."""
    if isinstance(info, text_type):
        info = info.encode("utf-8")
    digest_size = hashmod().digest_size
    N = int(math.ceil(L * 1.0 / digest_size))
    assert N <= 255
    T = b""
    output = []
    for i in xrange(1, N + 1):
        data = T + info + chr(i).encode("utf-8")
        T = hmac.new(PRK, data, hashmod).digest()
        output.append(T)
    return b"".join(output)[:L]


def HKDF(secret, salt, info, size, hashmod=hashlib.sha256):
    """HKDF-extract-and-expand as a single function."""
    PRK = HKDF_extract(salt, secret, hashmod)
    return HKDF_expand(PRK, info, size, hashmod)


# If httpie is installed, register the hawk plugin.
try:

    from httpie.plugins import AuthPlugin

    class HawkPlugin(AuthPlugin):

        name = 'Hawk Auth'
        auth_type = 'hawk'
        description = ''

        def get_auth(self, username, password):
            if password == '':
                return HawkAuth(hawk_session=username)
            return HawkAuth(id=username, key=password)

except ImportError:
    pass
